android/src/com/mapswithme/maps/MapFragment.java

```diff
@@ -2,140 +2,289 @@
+  {
+    nativeSetupWidget(WIDGET_COMPASS,
+                      UiUtils.dimen(R.dimen.margin_compass_left) + offset,
+                      mHeight - UiUtils.dimen(R.dimen.margin_compass_bottom),
+                      ANCHOR_CENTER);
+    if (forceRedraw)
+      nativeApplyWidgets();
+  }
 
-  protected native void nativeScalePlus();
-  protected native void nativeScaleMinus();
+  private void onRenderingInitialized()
+  {
+    final Activity activity = getActivity();
+    if (isAdded() && activity instanceof MapRenderingListener)
+      ((MapRenderingListener) activity).onRenderingInitialized();
+  }
 
-  public native boolean showMapForUrl(String url);
+  private void reportUnsupported()
+  {
+    new AlertDialog.Builder(getActivity())
+        .setMessage(getString(R.string.unsupported_phone))
+        .setCancelable(false)
+        .setPositiveButton(getString(R.string.close), new DialogInterface.OnClickListener()
+        {
+          @Override
+          public void onClick(DialogInterface dlg, int which)
+          {
+            getActivity().moveTaskToBack(true);
+          }
+        }).show();
+  }
 
   @Override
-  public void onCreate(Bundle savedInstanceState)
+  public void surfaceCreated(SurfaceHolder surfaceHolder)
   {
-    super.onCreate(savedInstanceState);
-    setRetainInstance(true);
+    final Surface surface = surfaceHolder.getSurface();
+    if (nativeIsEngineCreated())
+    {
+      nativeAttachSurface(surface);
+      mRequireResize = true;
+      return;
+    }
+
+    mRequireResize = false;
+    final Rect rect = surfaceHolder.getSurfaceFrame();
+    setupWidgets(rect.width(), rect.height());
+
+    final DisplayMetrics metrics = new DisplayMetrics();
+    getActivity().getWindowManager().getDefaultDisplay().getMetrics(metrics);
+
+    mEngineCreated = nativeCreateEngine(surface, metrics.densityDpi);
+    if (mEngineCreated)
+      onRenderingInitialized();
+    else
+      reportUnsupported();
   }
 
   @Override
-  public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState)
+  public void surfaceChanged(SurfaceHolder surfaceHolder, int format, int width, int height)
   {
-    return inflater.inflate(R.layout.fragment_map, container, false);
+    if (!mEngineCreated)
+      return;
+
+    if (!mRequireResize && surfaceHolder.isCreating())
+      return;
+
+    nativeSurfaceChanged(width, height);
+
+    mRequireResize = false;
+    setupWidgets(width, height);
+    nativeApplyWidgets();
   }
 
   @Override
-  public void onViewCreated(View view, @Nullable Bundle savedInstanceState)
+  public void surfaceDestroyed(SurfaceHolder surfaceHolder)
   {
-    super.onViewCreated(view, savedInstanceState);
-    nativeConnectDownloadButton();
+    if (!mEngineCreated)
+      return;
+
+    if (getActivity() == null || !getActivity().isChangingConfigurations())
+    {
+      nativeDestroyEngine();
+      MwmApplication.get().clearFunctorsOnUiThread();
```

Comment:
Definitely yes. Functors can contains pointer in DrapeEngine. We have to clean message queue before destroy engine